src/ResourceManager/Version2016_09_01/AzureRMCmdlet.cs

```diff
@@ -67,23 +70,66 @@ public IAzureContextContainer DefaultProfile
         {
             get
             {
-                if (_profile != null)
+                if (!ShouldCloneDefaultProfile())
                 {
-                    return _profile;
+                    return GetDefaultProfile();
                 }
-                if (AzureRmProfileProvider.Instance == null)
+
+                if (_clonedDefaultProfile == null)
                 {
-                    throw new InvalidOperationException(Resources.ProfileNotInitialized);
+                    _clonedDefaultProfile = CloneProfileAndModifyContext();
                 }
-
-                return AzureRmProfileProvider.Instance.Profile;
+                return _clonedDefaultProfile;
             }
             set
             {
                 _profile = value;
             }
         }
 
+        private IAzureContextContainer GetDefaultProfile()
+        {
+            if (_profile != null)
+            {
+                return _profile;
+            }
+            if (AzureRmProfileProvider.Instance == null)
+            {
+                throw new InvalidOperationException(Resources.ProfileNotInitialized);
+            }
+            return AzureRmProfileProvider.Instance.Profile;
+        }
+
+        private IAzureContextContainer CloneProfileAndModifyContext()
+        {
+            // going to modify default context, so only shallow copying other stuff
+            var clonedProfile = GetDefaultProfile().ShallowCopy();
+            clonedProfile.DefaultContext = clonedProfile.DefaultContext.DeepCopy();
+
+            if (MyInvocation.BoundParameters.TryGetValue(SubscriptionIdParameter, out var overriddenSub))
+            {
+                var matchingSub = clonedProfile.Subscriptions.FirstOrDefault(sub => sub.GetId().Equals(new Guid(overriddenSub as string)));
+                if (matchingSub != null)
```

Comment:
Should we tell customer the provided sub id is not found if matchingSub is null?